diff --git a/willuslib/willus.h b/willuslib/willus.h
--- a/willuslib/willus.h
+++ b/willuslib/willus.h
@@ -89,6 +89,10 @@
 
 typedef double  real;
 
+#if defined(__ANDROID__)
+#define ANDROID
+#endif
+
 #if (!defined(CYGWIN32) && defined(__CYGWIN32__))
 #define CYGWIN32
 #endif
@@ -170,7 +174,7 @@ typedef double  real;
 #define WILLUS_BIGENDIAN
 #endif
 
-#if (defined(__GNUC__) && __GNUC__ >= 4 && __GNUC_MINOR__ > 4)
+#if (defined(__GNUC__) && (__GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 5)))
 /* For my code */
 #ifndef WILLUS_HAVE_FILE64
 #define WILLUS_HAVE_FILE64
