diff --git a/willuslib/willus.h b/willuslib/willus.h
--- a/willuslib/willus.h
+++ b/willuslib/willus.h
@@ -257,6 +257,7 @@ typedef double  real;
 **
 */
 
+#if 0
 /*
 ** Determine if we can use winshell functions
 ** If CLANG version > 7 or GNU C version > 4
@@ -324,6 +325,9 @@ typedef double  real;
 #endif
 
 #endif /* USE_CMAKE */
+#else
+#include "willus_config.h"
+#endif
 /*
 ** Consistency check
 */
