diff --git a/src/wordrec/wordrec.h b/src/wordrec/wordrec.h
--- a/src/wordrec/wordrec.h
+++ b/src/wordrec/wordrec.h
@@ -19,9 +19,11 @@
 #ifndef TESSERACT_WORDREC_WORDREC_H_
 #define TESSERACT_WORDREC_WORDREC_H_
 
-#ifdef DISABLED_LEGACY_ENGINE
-
+#ifdef HAVE_CONFIG_H
 #include "config_auto.h"
+#endif
+
+#ifdef DISABLED_LEGACY_ENGINE
 
 #include <cstdint>             // for int16_t, int32_t
 #include "callcpp.h"           // for C_COL
