diff --git a/meson.build b/meson.build
--- a/meson.build
+++ b/meson.build
@@ -612,7 +612,6 @@ functions = [
   'close_range',
   'endmntent',
   'endservent',
-  'epoll_create',
   'fallocate',
   'fchmod',
   'fchown',
@@ -727,6 +726,10 @@ foreach f : functions
   endif
 endforeach
 
+if cc.has_function('epoll_create1')
+  glib_conf.set('HAVE_EPOLL_CREATE', 1)
+endif
+
 # Check that stpcpy() is usable; must use header.
 # See:
 # https://github.com/mesonbuild/meson/issues/5628.
