diff --git a/meson.build b/meson.build
--- a/meson.build
+++ b/meson.build
@@ -623,7 +623,6 @@ functions = [
   'copy_file_range',
   'endmntent',
   'endservent',
-  'epoll_create',
   'fallocate',
   'fchmod',
   'fchown',
@@ -743,6 +742,10 @@ foreach f : functions
   endif
 endforeach
 
+if cc.has_function('epoll_create1')
+  glib_conf.set('HAVE_EPOLL_CREATE', 1)
+endif
+
 # Export the information about free_sized() so we can correctly define a macro
 # wrapper around g_free()/g_free_sized() depending on whether it’s available
 glibconfig_conf.set('G_HAVE_FREE_SIZED', have_func_free_sized)
