diff --git a/gio/meson.build b/gio/meson.build
--- a/gio/meson.build
+++ b/gio/meson.build
@@ -173,6 +173,7 @@ gnetworking_h_conf.set('NAMESER_COMPAT_INCLUDE', gnetworking_h_nameser_compat_in
 gnetworking_h = configure_file(input : 'gnetworking.h.in',
                                output : 'gnetworking.h',
                                install_dir : gio_includedir,
+                               install_tag : 'devel',
                                configuration : gnetworking_h_conf)
 
 gdbus_headers = files(
