diff --git a/meson.build b/meson.build
--- a/meson.build
+++ b/meson.build
@@ -2120,8 +2120,7 @@ endif
 if libintl.found()
   have_bind_textdomain_codeset = cc.has_function('bind_textdomain_codeset', dependencies: libintl_deps)
 else
-  libintl = dependency('intl', allow_fallback: true)
-  assert(libintl.type_name() == 'internal')
+  libintl = subproject('proxy-libintl').get_variable('intl_dep')
   libintl_deps = [libintl]
   have_bind_textdomain_codeset = true  # proxy-libintl supports it
 endif
