--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -2,7 +2,7 @@
 #        you can define SDL3_INCLUDE_DIRS on the cmdline.  For example:
 #        cmake -DSDL3_INCLUDE_DIRS=/opt/SDL3/include [other stuff ....]
 
-cmake_minimum_required(VERSION 3.0...4.1)
+cmake_minimum_required(VERSION 3.17.5)
 project(sdl2_compat VERSION 2.32.60 LANGUAGES C)
 
 if(POLICY CMP0074)
@@ -611,8 +611,8 @@
   set(ENABLE_SHARED_TRUE "")
   set(ENABLE_SHARED_FALSE "#")
 
-  configure_file(sdl2-compat.pc.in sdl2-compat.pc @ONLY)
-  install(FILES "${CMAKE_CURRENT_BINARY_DIR}/sdl2-compat.pc"
+  configure_file(sdl2-compat.pc.in sdl2.pc @ONLY)
+  install(FILES "${CMAKE_CURRENT_BINARY_DIR}/sdl2.pc"
     DESTINATION ${CMAKE_INSTALL_LIBDIR}/pkgconfig
   )
 
