diff --git a/meson.build b/meson.build
--- a/meson.build
+++ b/meson.build
@@ -2,7 +2,7 @@ project(
   'libpng',
   'c',
   version: '1.6.43', license: 'libpng',
-  meson_version: '>=0.50.0',
+  meson_version: '>=0.60.0',
 )
 
 png_versions = meson.project_version().split('.')
@@ -129,6 +129,7 @@ configure_file(
   configuration: cdata,
   install_dir: get_option('libdir') / 'pkgconfig',
   install: true,
+  install_tag: 'devel',
 )
 
 # FIXME: this should be a symlink to libpng16.pc
@@ -138,6 +139,7 @@ configure_file(
   configuration: cdata,
   install_dir: get_option('libdir') / 'pkgconfig',
   install: true,
+  install_tag: 'devel',
 )
 
 libpng_dep = declare_dependency(
